src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -188,91 +192,147 @@ private static SequencePosition SeekMultiSegment(ReadOnlySequenceSegment<T> curr
+                }
+            }
+            else
+            {
+                // Multi-Segment Sequence
+                // Storing this in a local since it is used twice within InRange()
+                ulong startRange = (ulong)(((ReadOnlySequenceSegment<T>)startObject).RunningIndex + startIndex);
+                if (!InRange(
+                    (ulong)(((ReadOnlySequenceSegment<T>)start.GetObject()).RunningIndex + sliceStartIndex),
+                    startRange,
+                    (ulong)(((ReadOnlySequenceSegment<T>)endObject).RunningIndex + endIndex)))
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();
+                }
+            }
+        }
 
-                var startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
-                var endSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(endObject);
+        private void BoundsCheck(uint sliceStartIndex, object sliceStartObject, uint sliceEndIndex, object sliceEndObject)
+        {
+            uint startIndex = (uint)GetIndex(_sequenceStart);
+            uint endIndex = (uint)GetIndex(_sequenceEnd);
 
-                // start.RunningIndex + startIndex <= end.RunningIndex + endIndex
-                if (startSegment.RunningIndex - endIndex <= endSegment.RunningIndex - startIndex) // Rearranged to avoid overflow
+            object startObject = _sequenceStart.GetObject();
+            object endObject = _sequenceEnd.GetObject();
+
+            // Single-Segment Sequence
+            if (startObject == endObject)
+            {
+                if (sliceStartObject != sliceEndObject || sliceStartIndex > sliceEndIndex)
                 {
-                    // Mult-segment in bounds
-                    return;
+                    ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();
+                }
+
+                if (sliceStartIndex < startIndex || sliceEndIndex > endIndex)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();
+                }
+            }
+            else
+            {
+                // Multi-Segment Sequence
+                // This optimization works because we know sliceStartIndex, sliceEndIndex, startIndex, and endIndex are all >= 0
+                Debug.Assert(sliceStartIndex >= 0 && startIndex >= 0 && endIndex >= 0);
+
+                ulong sliceStartRange = (ulong)(((ReadOnlySequenceSegment<T>)sliceStartObject).RunningIndex + sliceStartIndex);
+                ulong sliceEndRange = (ulong)(((ReadOnlySequenceSegment<T>)sliceEndObject).RunningIndex + sliceEndIndex);
+
+                if (sliceStartRange > sliceEndRange)
+                    ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();
+
+                if (sliceStartRange < (ulong)(((ReadOnlySequenceSegment<T>)startObject).RunningIndex + startIndex)
```

Comment:
Add before
Debug.Assert(startObject != null);
Debug.Assert(endObject != null);

Reply:
On a well formed ReadOnlySequence, if startObject == null, then so is endObject which means it is single segment. We won't hit this case.

Reply:
Yes! But should we throw NullReference exception in debug mode or should add Assert?

Reply:
Since, it isn't absolutely necessary, I think we should address it separate from this PR.